pkgs/stdenv/generic/setup.sh

```diff
@@ -47,19 +51,67 @@ getAllOutputNames() {
     fi
 }
 
-# Logs arguments to $NIX_LOG_FD, if it exists, no-op if it does not.
-nixLog() {
-    if [[ -z ${NIX_LOG_FD-} ]]; then
-        return
-    fi
-    echo "$@" >&"$NIX_LOG_FD"
+# All provided arguments are joined with a space then directed to $NIX_LOG_FD, if it's set.
+# Corresponds to `Verbosity::lvlError` in the Nix source.
+nixErrorLog() {
+    if [[ -z ${NIX_LOG_FD-} ]] || [[ ${NIX_DEBUG:-0} -lt 0 ]]; then return; fi
+    printf "%s\n" "$*" >&"$NIX_LOG_FD"
+}
+
+# All provided arguments are joined with a space then directed to $NIX_LOG_FD, if it's set.
+# Corresponds to `Verbosity::lvlWarn` in the Nix source.
+nixWarnLog() {
+    if [[ -z ${NIX_LOG_FD-} ]] || [[ ${NIX_DEBUG:-0} -lt 1 ]]; then return; fi
+    printf "%s\n" "$*" >&"$NIX_LOG_FD"
+}
+
+# All provided arguments are joined with a space then directed to $NIX_LOG_FD, if it's set.
+# Corresponds to `Verbosity::lvlNotice` in the Nix source.
+nixNoticeLog() {
+    if [[ -z ${NIX_LOG_FD-} ]] || [[ ${NIX_DEBUG:-0} -lt 2 ]]; then return; fi
+    printf "%s\n" "$*" >&"$NIX_LOG_FD"
+}
+
+# All provided arguments are joined with a space then directed to $NIX_LOG_FD, if it's set.
+# Corresponds to `Verbosity::lvlInfo` in the Nix source.
+nixInfoLog() {
+    if [[ -z ${NIX_LOG_FD-} ]] || [[ ${NIX_DEBUG:-0} -lt 3 ]]; then return; fi
+    printf "%s\n" "$*" >&"$NIX_LOG_FD"
+}
+
+# All provided arguments are joined with a space then directed to $NIX_LOG_FD, if it's set.
+# Corresponds to `Verbosity::lvlTalkative` in the Nix source.
+nixTalkativeLog() {
+    if [[ -z ${NIX_LOG_FD-} ]] || [[ ${NIX_DEBUG:-0} -lt 4 ]]; then return; fi
+    printf "%s\n" "$*" >&"$NIX_LOG_FD"
+}
+
+# All provided arguments are joined with a space then directed to $NIX_LOG_FD, if it's set.
+# Corresponds to `Verbosity::lvlChatty` in the Nix source.
+nixChattyLog() {
+    if [[ -z ${NIX_LOG_FD-} ]] || [[ ${NIX_DEBUG:-0} -lt 5 ]]; then return; fi
+    printf "%s\n" "$*" >&"$NIX_LOG_FD"
+}
+
+# All provided arguments are joined with a space then directed to $NIX_LOG_FD, if it's set.
+# Corresponds to `Verbosity::lvlDebug` in the Nix source.
+nixDebugLog() {
+    if [[ -z ${NIX_LOG_FD-} ]] || [[ ${NIX_DEBUG:-0} -lt 6 ]]; then return; fi
+    printf "%s\n" "$*" >&"$NIX_LOG_FD"
+}
+
+# All provided arguments are joined with a space then directed to $NIX_LOG_FD, if it's set.
+# Corresponds to `Verbosity::lvlVomit` in the Nix source.
+nixVomitLog() {
+    if [[ -z ${NIX_LOG_FD-} ]] || [[ ${NIX_DEBUG:-0} -lt 7 ]]; then return; fi
+    printf "%s\n" "$*" >&"$NIX_LOG_FD"
```

Comment:
Can we avoid duplicating the implementation if we add log levels as function parameters? What do you think about passing human-readable log level? E.g.

```bash
declare -r -A _nixLogLevels=(
	[error]=0
	[warn]=1
	[notice]=2
	[info]=3
	[talkative]=4
	[chatty]=5
	[debug]=6
	[vomit]=7
)

_nixLog() {
	local logLevelName=$1
	shift
	local logMessage=$*

	local logLevelIndex=${_nixLogLevels[$logLevelName]}

	if [[ -z ${NIX_LOG_FD-} ]]; then
		return
	fi

	# Temporarily use NIX_DEBUG to suppress logs until we figure out how to
	# properly integrate with Nix structured logs.
	if [[ ${NIX_DEBUG:-0} -lt $logLevelIndex ]]; then
		return
	fi

	printf "%s\n" "$logMessage" >&"$NIX_LOG_FD"
}
```

Reply:
I don't like the proposed de-duplication very much. What I like about the bullheaded duplication is that it's damned simple, taggable, and documentable. Since it's two lines of actual implementation -- with a condition on `NIX_DEBUG` that I hope disappears -- I'm going to keep it this way.

Reply:
Yeah, `_nixLogLevels` is a bit overkill… Perhaps
```bash
# Print all arguments to the file descriptor specified in $NIX_LOG_FD if set,
# otherwise the function is no-op.
_nixLogToFd() {
    if [[ -z ${NIX_LOG_FD-} ]]; then
        return
    fi
    printf "%s\n" "$*" >&"$NIX_LOG_FD"
}

# All provided arguments are joined with a space then directed to $NIX_LOG_FD,
# if it’s set. Corresponds to `Verbosity::lvlError` in the Nix source.
nixErrorLog() {
    [[ ${NIX_DEBUG:-0} -ge 1 ]] || return 0
    _nixLogToFd "$*"
}

# All provided arguments are joined with a space then directed to $NIX_LOG_FD,
# if it’s set. Corresponds to `Verbosity::lvlWarn` in the Nix source.
nixWarnLog() {
    [[ ${NIX_DEBUG:-0} -ge 1 ]] || return 0
    _nixLogToFd "$*"
}

# …
```

This way, we can use `_nixLogToFd` for `setPhase` and other higher-level structured log output that should go to `NIX_LOG_FD`.

Feel free to mark this conversation as resolved if it doesn’t click with you as well.

Reply:
I'm planning on doing more work in this area, and I'll leave it for then, especially since I'll be experimenting with structured logging.